---
title: "assignment"
author: "Catherine Rohner"
date: "2024-05-14"
output: html_document
---

## Load the Packages

```{r}
library(GenomicRanges)
library(rtracklayer)
library(epiwraps)
library(ggplot2)
library(rGREAT) # Gene Ontology enrichment among genomic regions
```

## Download the Data
```{r}
setwd("/Users/catherinerohner/ETH/Bioinformatics/Assignment10")
download.file("https://ethz-ins.org/content/w10.assignment.zip", "w10.assignment.zip")
unzip("w10.assignment.zip")
list.files()
```
## Filter and import high-confidence peaks to define regions for analysis. Load corresponding signal tracks.

```{r}
peaks <- list.files(pattern="bed$")
# we first import the peaks
peaks <- lapply(peaks, rtracklayer::import.bed)
# we'll focus on the high-quality peaks
peaks <- lapply(peaks, FUN=function(x) x[x$score>800])
# we get the union of non-redundant regions
regions <- reduce(unlist(GRangesList(peaks)))
#l load tracks
tracks <- list.files(pattern="bw$")
```

## Analyze Relationships Between CREB Family Transcription Factors

We utilize enriched heatmaps to examine the binding patterns of CREB1, CREB3, and CREB3L1 across the genomic regions. The initial plots reveal clear signals for all three transcription factors, suggesting strong binding activity. However, the complexity and overlap in data require further organization to better identify and understand the underlying patterns

```{r}
ese <- signal2Matrix(tracks, regions, extend=2000)
plotEnrichedHeatmaps(ese, use_raster=FALSE)
```

## Find the appropriate number of clusters by analyzing the explained variance for each cluster count from 2 to 10.

The graph illustrates that k-values of 3, 4, and 5 are potential choices for clustering. Based on our discussions in class, it is often advantageous to start with a higher number of clusters (k=5) and then consolidate them if they appear redundant.

```{r}
cl2 <- clusterSignalMatrices(ese, k=2:10)
ggplot(cl2$varExplained, aes(k, varExplained)) + geom_line()
```


## Initial Clustering Analysis

We initially segmented the data into five clusters, which explained 74% of the variance. A comparison with four clusters, which accounted for 70% of the variance, shows minimal improvement in variance explanation when increasing to five clusters. This marginal gain suggests that clustering into four groups might be more appropriate. Additionally, clusters 2 and 3 in the five-cluster model contain notably fewer data points, which could indicate overfitting. This observation warrants further analysis to confirm the optimal cluster count.

```{r}
set.seed(123)  # to ensure that it gives the same results everytime
cl <- clusterSignalMatrices(ese, k=5)
table(cl)
head(cl)
length(cl)
length(regions)

rowData(ese)$cluster <- cl
```
## Plotting Clustered Data with Enriched Heatmaps 

Upon visualizing the data, Clusters 2, 4 and 5 appear remarkably similar, underscoring our previous observations about diminishing returns when adding more clusters. In line with the strategy discussed in class to merge similar clusters for a clearer and more meaningful analysis, I will proceed by reducing the number of clusters to four. 

```{r}
mycolors <- c("1"="red", "2"="blue", "3"="darkgreen", "4"="black", "5" = "yellow")
plotEnrichedHeatmaps(ese, row_split="cluster",trim=0.99, mean_color=mycolors, colors=c("white","darkred"),use_raster=FALSE)
```

## Detailed Clustering for k=4

The heatmap distinctly showcases four clusters, validating the decision to reduce the cluster count. Observations from the heatmap are as follows:

Cluster 1: Demonstrates the highest activity for Creb3L1, which is significantly supported by substantial signals from both Creb1 and Creb3. This diverse range of active transcription factors suggests a versatile regulatory region with multifunctional capabilities.

Cluster 2: Predominantly characterized by a strong signal for Creb1, with moderate expression of Creb3L1 and only minimal activity from Creb3. This pattern suggests that Cluster 2 has a specific functional emphasis, primarily driven by Creb1.

Cluster 3: Shows a balanced, high activity for both Creb1 and Creb3L1, but with very little presence of Creb3. This configuration indicates a unique regulatory interaction that differs from the other clusters, possibly involving synergistic effects between Creb1 and Creb3L1.

Cluster 4: Marked by a notably high signal for Creb3, setting it apart from the other clusters. This distinct profile underscores Creb3's dominant role within this cluster, potentially indicating specialized regulatory functions exclusive to Creb3.

```{r}
cl <- clusterSignalMatrices(ese, k=4)
rowData(ese)$cluster <- cl
table(cl)
head(cl)
mycolors <- c("1"="red", "2"="blue", "3"="darkgreen", "4"="black")
plotEnrichedHeatmaps(ese, row_split="cluster", mean_color=mycolors, colors=c("white","darkred"),use_raster=FALSE)
```

## Visualizing the average signals across the clusters allows us to confirm specific characteristics:

Cluster 1: Displays a balanced level of activity between Creb3L1 and Creb1 and Creb3, indicating their co-regulation in this genomic region.

Cluster 2: Shows a distinct, strong signal for Creb1, highlighting its dominant role and specialized activity within this cluster.

Cluster 3: Initially appears similar to Cluster 1 but upon closer inspection, it shows a notably lower peak for Creb3 while maintaining more pronounced signals for Creb1 and Creb3L1. This suggests a unique and complex regulatory pattern, differentiating it from Cluster 1.

Cluster 4: Characterized by a high peak for Creb3, emphasizing its significant influence and specialized function in this cluster.

```{r}
d <- meltSignals(ese, splitBy=cl)
rowData(ese)$cluster <- cl
ggplot(d, aes(position, mean, colour=sample)) + geom_line(size=1.2) + facet_wrap(~split)
```

## Enhancing Feature Comparison with Relative Scaling

Cluster 1: Displays high peaks for Creb3L1 accompanied by moderate signals for Creb1, suggesting both play significant roles in this cluster. Notably, Creb3, which was more visible in the unscaled data, is barely detectable here.

Cluster 2: Dominated by a strong signal for Creb3, with minimal activity from other transcription factors. This is a shift from the previous heatmap where Creb1 was predominant. The change from Creb1 to Creb3 dominance in the scaled heatmap suggests that although Creb3 might be weaker in absolute terms, it exhibits relatively stronger or more consistent activity within this cluster when scaled.

Cluster 3: Characterized by very high peaks for Creb1, underscoring its dominant influence in this clusterâ€”a stark contrast to the other transcription factors. The scaling emphasizes Creb1's impact, overshadowing Creb3L1 and rendering Creb3 nearly invisible, suggesting that Creb1's role is considerably amplified under relative scaling.

Cluster 4: Shows prominent activity for Creb3L1, indicating its significant regulatory role, in contrast to its previous dominance by Creb3. This shift suggests that Creb3, while showing spikes of high activity in absolute terms that skew its perceived importance, exhibits less consistent activity than Creb3L1 when signals are scaled relatively.

These discrepancies are explainable due to the use of different methodologies in signal scaling and normalization. Absolute scaling, which presents raw signal intensities, can mask subtle interplays between transcription factors. In contrast, relative scaling highlights differences in expression patterns across clusters by normalizing these signals, revealing unique regulatory dynamics that might not be apparent with absolute values alone.

```{r}
cl <- clusterSignalMatrices(ese, k=4, scaleRows = TRUE)
d <- meltSignals(ese, splitBy=cl)
ggplot(d, aes(position, mean, colour=sample)) + geom_line() + facet_wrap(~split)
mycolors <- c("1"="red", "2"="blue", "3"="darkgreen", "4"="black")
plotEnrichedHeatmaps(ese, row_split = cl,mean_color=mycolors, scale_rows = "global",use_raster=FALSE)
```

## Get the description of the Cluster1.

Looking at the values, we can see that the enrichment analysis highlights a significant involvement of genomic regions in processes related to cell communication and response to stimuli, including "cell communication," "cellular response to stimulus," "signaling," "response to stimulus," "signal transduction," and "response to chemical." These processes show notable enrichment with observed region hits ranging from 180 to 311, indicating a strong representation within the analyzed genomic regions. Specifically, "cell communication" and "response to stimulus" are among the most enriched processes, with 270 and 311 hits respectively, suggesting these areas as key functions regulated by the genomic regions in this cluster. Fold enrichment values range from about 1.18 to 1.32, underscoring a moderate overrepresentation of these terms relative to the background. The hyper-p-values are highly significant, ranging from approximately 7.24e-07 to 0.0098, reinforcing the statistical significance of the enrichment.

Cluster 4 shows prominent activity for Creb3L1, indicating its significant regulatory role, in contrast to its previous dominance by Creb3. This shift suggests that Creb3, while showing spikes of high activity in absolute terms, exhibits less consistent activity than Creb3L1 when signals are scaled relatively. The enriched processes related to cell communication and response to stimuli align well with Creb3L1's regulatory role, suggesting that Creb3L1 is crucial in coordinating cellular interactions and responses to environmental changes. This consistent activity of Creb3L1 likely drives the regulatory mechanisms within these enriched processes, highlighting its importance in maintaining cellular homeostasis and responding to external signals.

Creb3 may have shown higher activity in absolute terms due to its role in rapid and dynamic responses to stress and other stimuli, leading to spikes in activity. This characteristic makes Creb3 highly responsive under specific conditions but less consistent overall compared to Creb3L1. 

```{r}
split_regions <- split(rowRanges(ese), rowData(ese)$cluster)

res <- great(split_regions[["4"]], gene_sets="GO:BP", tss_source="hg38", background=regions, cores=2)
bp <- getEnrichmentTables(res)
head(bp)
```
## Use ggplot2 to visualize results from enrichment analysis

The plot reveals significant enrichment of processes related to cell communication and response to stimuli, as evidenced by their high fold enrichment and low adjusted p-values. Processes such as "multicellular organismal process," "regulation of cellular process," and "regulation of biological process" involve a large number of genomic regions but are less statistically significant. This indicates that while these processes are common in the dataset, their enrichment is not as specific or strong compared to processes like "cell communication" and "cellular response to stimulus," which are crucial for coordinating cellular interactions and responses to environmental changes.

```{r}
ggplot(head(bp, 15), aes(fold_enrichment, reorder(description, p_adjust), size=observed_region_hits, color=-log10(p_adjust))) +
  geom_point() + scale_color_viridis_c()
```

